qiskit/version.py

```diff
@@ -81,35 +82,70 @@ def get_version_info():
 __version__ = get_version_info()
 
 
-def _get_qiskit_versions():
-    out_dict = {}
-    out_dict['qiskit-terra'] = __version__
-    try:
-        from qiskit.providers import aer
-        out_dict['qiskit-aer'] = aer.__version__
-    except Exception:
-        out_dict['qiskit-aer'] = None
-    try:
-        from qiskit import ignis
-        out_dict['qiskit-ignis'] = ignis.__version__
-    except Exception:
-        out_dict['qiskit-ignis'] = None
-    try:
-        from qiskit.providers import ibmq
-        out_dict['qiskit-ibmq-provider'] = ibmq.__version__
-    except Exception:
-        out_dict['qiskit-ibmq-provider'] = None
-    try:
-        from qiskit import aqua
-        out_dict['qiskit-aqua'] = aqua.__version__
-    except Exception:
-        out_dict['qiskit-aqua'] = None
-    try:
-        out_dict['qiskit'] = pkg_resources.get_distribution('qiskit').version
-    except Exception:
-        out_dict['qiskit'] = None
-
-    return out_dict
-
-
-__qiskit_version__ = _get_qiskit_versions()
+class QiskitVersion(Mapping):
+    """A lazy loading wrapper to get qiskit versions."""
+
+    __slots__ = ['_version_dict', '_loaded']
+
+    def __init__(self):
+        self._version_dict = {
+            'qiskit-terra': __version__,
```

Comment:
Should the `get_version_info()` call that populates `__version__` also be moved to runtime (or maybe this needs python 3.7)?

Reply:
Ideally yes it would be good to do this only at runtime, although for the release case (or anything outside of the dev case) this only adds an extra `stat()` call which is limited overhead.  But I think we'll have to wait until python 3.7 with module `getattr` to do this.

We could create a custom class with a `__str__` and `__repr__` that runs `get_version_info()` internally but that would cause issues for people who use string methods on version (the use case I'm thinking of is `qiskit.__version__.split('.')` so I think it would be best to wait